file.go

```diff
@@ -57,45 +57,117 @@ func WriteFile(path string, data string) error {
-		fb.fileNames = append(fb.fileNames, path)
+	m := &gitignoreFileMatcher{
+		ignorer: ignorer,
+		prefix:  prefix,
 	}
-	return nil
+	return m, nil
 }
 
-// BuildFileList returns a list of files that match the given list of expressions
-// rooted at the given startPath. The expressions correspond to gitignore ignore
-// expressions: anything that would be matched - and therefore ignored by git - is included
-// in the returned list of file paths. BuildFileList does not follow symlinks as
-// it uses filpath.Walk, which does not follow symlinks.
-func BuildFileList(startPath string, expressions ...string) ([]string, error) {
-	ignorer, err := ignore.CompileIgnoreLines(expressions...)
-	if err != nil {
-		return nil, err
+func (m *gitignoreFileMatcher) Match(file string, info os.FileInfo) bool {
+	file = strings.TrimLeft(strings.TrimPrefix(file, m.prefix), string(os.PathSeparator))
+	return !info.IsDir() && m.ignorer.MatchesPath(file)
+}
+
+// FileMatcher represents a type that can match against files and file
+// information to determine if it should be included.
+type FileMatcher interface {
+	Match(file string, info os.FileInfo) bool
+}
+
+// AlwaysMatch provides a FileMatcher implementation that will always match a
+// file or directory.
+type AlwaysMatch struct{}
+
+// Match always returns true.
+func (m AlwaysMatch) Match(string, os.FileInfo) bool { return true }
+
+// FileFileAlwaysMatch provides a FileMatcher implementation that will always
+// match a file, but not a directory.
+type FileAlwaysMatch struct{}
+
+func (m FileAlwaysMatch) Match(_ string, info os.FileInfo) bool {
+	return !info.IsDir()
+}
+
+// NeverMatch provides a FileMatcher implementation that will never match a
+// file or directory.
+type NeverMatch struct{}
+
+// Match always returns false.
+func (m NeverMatch) Match(string, os.FileInfo) bool { return false }
+
+// FileListBuilder provides options to find files within a directory.
+type FileListBuilder struct {
+	// Include determines which files should be included. This has lower
+	// precedence than the Exclude filter.
+	Include FileMatcher
+	// Exclude determines which files should be excluded from the file list.
+	// This has higher precedence than the Include filter.
+	Exclude FileMatcher
+	// WorkingDir is the base working directory from which to start searching
+	// for files.
+	WorkingDir string
+	files      []string
+}
+
+// NewFileListBuilder returns a default FileListBuilder that will match all
+// files (but not directory names) within the given directory dir.
+func NewFileListBuilder(dir string) *FileListBuilder {
+	return &FileListBuilder{
+		Include:    FileAlwaysMatch{},
+		Exclude:    NeverMatch{},
+		WorkingDir: dir,
+	}
+}
+
+// Build finds all files that pass the include and exclude filters within the
+// working directory. It does not follow symlinks. All files returned are
+// relative to the working directory.
+func (b *FileListBuilder) Build() ([]string, error) {
+	if b.Include == nil {
```

Comment:
Why does their have to be an include filter?

Reply:
Because if you don't specify what files you will include, by default, it includes nothing. You should use `FileAlwaysInclude` or `AlwaysInclude` to determine the behavior.